src/xenia/apu/audio_decoder.cpp

```diff
@@ -0,0 +1,239 @@
+/**
+******************************************************************************
+* Xenia : Xbox 360 Emulator Research Project                                 *
+******************************************************************************
+* Copyright 2015 Ben Vanik. All rights reserved.                             *
+* Released under the BSD license - see LICENSE in the root for more details. *
+******************************************************************************
+*/
+
+#include "xenia/apu/audio_decoder.h"
+
+#include "xenia/apu/audio_system.h"
+#include "xenia/base/logging.h"
+
+extern "C" {
+  #include "libavcodec/avcodec.h"
+}
+
+// Credits for most of this code goes to:
+// https://github.com/koolkdev/libertyv/blob/master/libav_wrapper/xma2dec.c
+
+namespace xe {
+namespace apu {
+
+AudioDecoder::AudioDecoder() : offset_(0) {
+}
+
+AudioDecoder::~AudioDecoder() {
+  if (context_) {
+    if (context_->extradata) {
+      delete context_->extradata;
+    }
+    if (avcodec_is_open(context_)) {
+      avcodec_close(context_);
+    }
+    av_free(context_);
+  }
+  if (decoded_frame_) {
+    avcodec_free_frame(&decoded_frame_);
+  }
+  if (current_frame_) {
+    delete current_frame_;
+  }
+}
+
+int AudioDecoder::Initialize(int bits) {
+  static bool avcodec_initialized = false;
+  if (!avcodec_initialized) {
+    avcodec_register_all();
+    avcodec_initialized = true;
+  }
+
+  if (bits <= 0 || bits > 32 || (bits % 8) != 0) {
+    assert_always();
+  }
+
+  // Output bits per sample
+  bits_ = bits;
+
+  // Allocate important stuff
+  codec_ = avcodec_find_decoder(AV_CODEC_ID_WMAPRO);
```

Comment:
it's probably worth logging on these failures (XELOGE) - unless they're super unlikely